tests/src/tests/component_test_utils.go

```diff
@@ -0,0 +1,228 @@
+	}
+
+	if GetBinaryDecision() {
+		containerComponent.VolumeMounts = AddVolume(GetRandomNumber(4))
+	} else {
+		containerComponent.Env = nil
+	}
+
+	if GetBinaryDecision() {
+		containerComponent.Endpoints = CreateEndpoints()
+	}
+
+}
+
+func setVolumeComponentValues(volumeComponent *schema.VolumeComponent) {
+
+	if GetRandomDecision(5, 1) {
+		volumeComponent.Size = strconv.Itoa(4+GetRandomNumber(252)) + "G"
+		LogMessage(fmt.Sprintf("   ....... volumeComponent.Size: %s", volumeComponent.Size))
+	}
+
+}
+
+func (devfile *TestDevfile) UpdateComponent(component *schema.Component) error {
+
+	errorString := ""
+	testComponent, found := getSchemaComponent(devfile.SchemaDevFile.Components, component.Name)
+	if found {
+		LogMessage(fmt.Sprintf(" ....... Updating component name: %s", component.Name))
+		if testComponent.ComponentType == schema.ContainerComponentType {
+			setContainerComponentValues(component.Container)
+		} else if testComponent.ComponentType == schema.VolumeComponentType {
+			setVolumeComponentValues(component.Volume)
+		}
+	} else {
+		errorString += LogMessage(fmt.Sprintf(" ....... Component not found in test : %s", component.Name))
+	}
+	var err error
+	if errorString != "" {
+		err = errors.New(errorString)
+	}
+	return err
+}
+
+func (devfile TestDevfile) VerifyComponents(parserComponents []schema.Component) error {
+
+	LogMessage("Enter VerifyComponents")
+	errorString := ""
+
+	// Compare entire array of commands
+	if !cmp.Equal(parserComponents, devfile.SchemaDevFile.Components) {
+		errorString += LogMessage(fmt.Sprintf(" --> ERROR: Component array compare failed."))
+		for _, component := range parserComponents {
+			if testComponent, found := getSchemaComponent(devfile.SchemaDevFile.Components, component.Name); found {
+				if !cmp.Equal(component, *testComponent) {
+					parserFilename := AddSuffixToFileName(devfile.FileName, "_"+component.Name+"_Parser")
+					testFilename := AddSuffixToFileName(devfile.FileName, "_"+component.Name+"_Test")
+					LogMessage(fmt.Sprintf("   .......marshall and write devfile %s", parserFilename))
+					c, err := yaml.Marshal(component)
+					if err == nil {
+						err = ioutil.WriteFile(parserFilename, c, 0644)
+					}
+					LogMessage(fmt.Sprintf("   .......marshall and write devfile %s", testFilename))
+					c, err = yaml.Marshal(testComponent)
+					if err == nil {
+						err = ioutil.WriteFile(testFilename, c, 0644)
+					}
+					errorString += LogMessage(fmt.Sprintf(" --> ERROR: Component %s did not match, see files : %s and %s", component.Name, parserFilename, testFilename))
+				} else {
+					LogMessage(fmt.Sprintf(" --> Component matched : %s", component.Name))
+				}
+			} else {
+				errorString += LogMessage(fmt.Sprintf(" --> ERROR: Component from parser not known to test - id : %s ", component.Name))
+			}
+		}
+		for _, component := range devfile.SchemaDevFile.Components {
+			if _, found := getSchemaComponent(parserComponents, component.Name); !found {
```

Comment:
since `parserComponents` and `devfile.SchemaDevFile.Components` are already compared using `cmp.Equal`, this for loop is redundant

Reply:
rechecked,  the loop is executed if the two slices do not match. my bad

Reply:
The cmp.Equal doesn't give an error if it fails, so you don't which component(s) was in error. The second loop only runs if the cmp fails and it does a component by component compare so it can narrow down the error to a specific component. Just helps with diagnosis if an error should happen.